### ISSUE TITLE
ansible-lint errors when using 2.7+ include_tasks apply argument

### ISSUE THREAD

**Issue by zoredache** [ansible/ansible-lint#507]
# Issue Type
- Bug report

# Ansible and Ansible Lint details

    $ ansible --version
    ansible 2.7.10
      config file = /etc/ansible/ansible.cfg
      configured module search path = ['/home/ansible/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
      ansible python module location = /usr/local/lib/python3.7/dist-packages/ansible
      executable location = /usr/local/bin/ansible
      python version = 3.7.3rc1 (default, Mar 13 2019, 11:01:15) [GCC 8.3.0]

    $ ansible-lint --version
    ansible-lint 4.1.0

- installation method: `pip3 install ansible ansible-lint` inside debian:buster container from docker hub

# Desired Behaviour

Shouldn't fail on valid syntax for 2.7+

- https://docs.ansible.com/ansible/latest/modules/include_tasks_module.html#parameters

# Actual Behaviour (Bug report only)

Example Files

    $ more *
    ::::::::::::::
    playbook.yml
    ::::::::::::::
    - hosts: localhost
      gather_facts: no
      tasks:
      - name: simple include task with apply
        include_tasks:
          file: debug.yml
          apply:
            tags:
              - debug
    ::::::::::::::
    debug.yml
    ::::::::::::::
    - debug:

    ::::::::::::::
    roles/example/main.yml
    ::::::::::::::
    - include_tasks:
        file: debug.yml
        apply:
          tags:
          - debug
    ::::::::::::::
    roles/example/debug.yml
    ::::::::::::::
    - debug:

Errors from ansible-lint 

    $ ansible-lint roles/example/
    WARNING: Couldn't open /home/ansible/project/roles/example/tasks/{'file': - No such file or directory

    $ ansible-lint playbook.yml 
    Traceback (most recent call last):
      File "/usr/local/bin/ansible-lint", line 10, in <module>
        sys.exit(main())
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/__main__.py", line 187, in main
        matches.extend(runner.run())
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/__init__.py", line 264, in run
        for child in ansiblelint.utils.find_children(arg, self.playbook_dir):
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 171, in find_children
        for child in play_children(basedir, item, playbook[1], playbook_dir):
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 223, in play_children
        return delegate_map[k](basedir, k, v, parent_type)
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 243, in _taskshandlers_children
        append_children(th['include_tasks'], basedir, k, parent_type, results)
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 273, in append_children
        'path': path_dwim(basedir, taskhandler),
      File "/usr/local/lib/python3.7/dist-packages/ansiblelint/utils.py", line 72, in path_dwim
        return dl.path_dwim(given)
      File "/usr/local/lib/python3.7/dist-packages/ansible/parsing/dataloader.py", line 181, in path_dwim
        given = unquote(given)
      File "/usr/local/lib/python3.7/dist-packages/ansible/parsing/quoting.py", line 29, in unquote
        if is_quoted(data):
      File "/usr/local/lib/python3.7/dist-packages/ansible/parsing/quoting.py", line 24, in is_quoted
        return len(data) > 1 and data[0] == data[-1] and data[0] in ('"', "'") and data[-2] != '\\'
    KeyError: 0

Yet these are a valid playbook and role

    $ ansible-playbook -i localhost, playbook.yml 

    PLAY [localhost] ******************************************************************************************************

    TASK [simple include task with apply] *********************************************************************************
    included: /home/ansible/project/debug.yml for localhost

    TASK [debug] **********************************************************************************************************
    ok: [localhost] => {
        "msg": "Hello world!"
    }

    PLAY RECAP ************************************************************************************************************
    localhost  

    $ ansible localhost -i localhost, -m include_role -a name=roles/example
    localhost | SUCCESS => {
        "changed": false,
        "include_variables": {
            "name": "roles/example"
        }
    }
    localhost | SUCCESS => {
        "changed": false,
        "include": "debug.yml",
        "include_variables": {
            "apply": {
                "tags": [
                    "debug"
                ]
            }
        }
    }
    localhost | SUCCESS => {
        "msg": "Hello world!"
    }

**Comment by carpioldc** [comment]
Same problem here. Any updates?

**Comment by hobbestigrou** [comment]
I have the same error. I have made a patch in ansible that fix this error:

https://github.com/ansible/ansible/pull/58725

I'm going also make a patch in ansible-lint for don't call this ansibe function if the data is not a string.

**Comment by rnsc** [comment]
I have the same issue with include_tasks with the new approach using apply.
It doesn't throw errors if I use free-form and args: though.

**Comment by hobbestigrou** [comment]
Please @rnsc can you paste the traceback? I suppose the error is the same, in this case like I said previously I have made a patch in ansible, that fix this error. But effectively I must also modify the linter.

**Comment by rnsc** [comment]
@hobbestigrou I don't any traceback:

`ansible-lint -vvv tasks/main.yml
Examining tasks/main.yml of type playbook
Examining /redacted/path/tasks/{u'apply': of type playbook
WARNING: Couldn't open /redacted/path/tasks/{u'apply': - No such file or directory`

This is what the file looks like:
`- name: "include_tasks | {{ role_path | basename }}"
  include_tasks:
    file: "{{ policy_item }}"
    apply:
      become: true
  loop_control:
    loop_var: policy_item
  with_items:
    - "{{ lookup('fileglob',  role_path + '/tasks/task_*.yml').split(',') | sort }}"`

**Comment by rnsc** [comment]
FYI, I'm just using ansible-lint 4.1.0 as a pip package.

**Comment by hobbestigrou** [comment]
Thanks @rnsc, it's weird do you have no traceback. Anyway, can you please tried to apply the patch I have propose to ansible? Apparently it's the same error.

**Comment by rnsc** [comment]
@hobbestigrou I overwrote my local quoting.py copy and removed quoting.pyc in my Virtual Env, and I still have that error. I'm failing to see how that patch is linked to this though. But maybe I'm missing information about how ansible-lint works.

**Comment by hobbestigrou** [comment]
It's linked @rnsc because the linter call this function. Read the traceback from the first comment to see this.

**Comment by rnsc** [comment]
> It's linked @rnsc because the linter call this function. Read the traceback from the first comment to see this.

Sure, but why wouldn't it work for me then?
Were you able to test this with your patch?

**Comment by hobbestigrou** [comment]
I don't know why @rnsc that doesn't works for you. I don't know how you test that. For example, do you use virtualenv, if yes, are your sure you have modify the file in the virtualenv? How do you have apply the patch? Yes I have test the patch and for me that works with the initial bug.

**Comment by rnsc** [comment]
@hobbestigrou thank you for your time, I'm going to test again and override my parsing.py in my virtualenv.

**Comment by rnsc** [comment]
@hobbestigrou nope, doesn't work. What version of Ansible are you using?
I'm using Ansible 2.8.2.
I overwrote `~/VE282/lib/python2.7/site-packages/ansible/parsing/quoting.py` and deleted the `quoting.pyc`.
After running ansible-lint, I still have the error and I can see the `quoting.pyc` generated, meaning that it recompiled it. Unless I'm mistaken.

**Comment by hobbestigrou** [comment]
You're welcome @rnsc for the time. It's really weird, but it's hard to help you without a traceback. The version of ansible is not important in this case.

**Comment by rnsc** [comment]
@hobbestigrou even with -vvv, ansible-lint doesn't give much details about what is wrong.
The funny thing is, it doesn't prevent the linting from happening. It just throws a warning. As if it was trying to resolve a file on the filesystem somewhere based on the content of the YAML.

**Comment by hobbestigrou** [comment]
You have not a KeyError exception? Check the first comment for example.

**Comment by rnsc** [comment]
@hobbestigrou well, the first comment is depicting two issues.
One that throws a warning and one that throws an exception. I only have the warning when I test.

**Comment by hobbestigrou** [comment]
Sorry @rnsc I have misunderstood, the patch I have made in ansible fix only the exception. I'm going look for fix the warning today.

**Comment by gwsales** [comment]
Nearly a year old, but I'm also having this issue. Anyone have an example of the args workaround?

**Comment by Nethe** [comment]
@gwsales 
> Nearly a year old, but I'm also having this issue. Anyone have an example of the args workaround?

Hey. I found some workaround, but it is not working to me. Please see [THIS LINK](https://www.gitmemory.com/issue/ansible/molecule/2171/510901125)

**Comment by wedi** [comment]
@gwsales The workaround works like this. Write

```yaml
- include_tasks: my_tasks.yml
  args:
    apply:
      tags: ['software_installation']
```

instead of

```yaml
- include_tasks:
  file: my_tasks.yml
    apply:
      tags: ['software_installation']
```

Personally, I don't consider a warning triggered by a bug that does not affect the lint result is worth a change although my inner Monk revolts.

**Comment by ssbarnea** [comment]
I am unable to reproduce this bug with current version. If you can reproduce it with code from master, please raise new bug, one that contains the minimal example possible to reproduce the failure.

**Comment by zoredache** [comment]
@ssbarnea

I disagree about this being 'fixed'. There are two separate cases described above. In the first case the error is hidden, but ansible-lint doesn't actually work properly. The second case is not fixed at all and continues to fail.

The **two** cases are 

- ansible playbook that has an include_tasks with an apply
- ansible role that has an include_tasks with an apply

So with the latest dev build you do not get an error when running ansible-lint against the playbook that includes a tasks file, but **it doesn't actually validate** the included tasks file.

With the latest dev build and a role with a include, it continues to fail with you lint the role directly or the indirectly.

I have just added a demonstration and output I am getting in a new git repository. Maybe the problem here is that I didn't communicate the problem well enough. The output is in the README.md

https://github.com/zoredache/ansible_lint_error_demonstration

So if you review my output, the things will want to note is that the output for `ansible-lint -vvv playbook_inc_tasks.yml` for the files that have an apply.  When you have an apply, the included file is not examined `DEBUG    Examining debug.yml of type tasks`.

The case of an apply in a role fails when examining a role directly `ansible-lint -vvv roles/example` or indirectly `ansible-lint -vvv playbook_inc_role.yml`. See the `Traceback` output for both.
